### DISCUSSION TITLE
shell: portable scripts

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#12138] [Ideas]
## Problem

One benefit of `dagger shell` is that it could potentially match the *convenience* of bash scripts. 👍 One drawback is that it could potentially match the *portability* of bash scripts 👎 

Since the shell interprets scripts *in the CLI*, it gets its own sandbox, separate from module runtimes. This is an opportunity to design the sandbox for the ideal balance of portability and convenience.

## Solution

We are still searching for the ideal balance of portability and convenience. Let's use this issue for discussion.

Here is a summary of relevant features, and how they are sandboxed in each tool.

- ❌ means "not available"
- 😈 means "available but unsafe to share"
- 😇 means "available and safe to share"

| | Bash script | `dagger call` | `dagger shell` v0.14.0 | Dagger module |
| -- | -- | -- | -- | -- |
| Execute command on the host | 😈  | 😈   `--token cmd:aws-vault` | ❌ | ❌ |
| Read host filesystem | 😈  | 😈  `--source /Users/solomon/my/src` | 😈  `.host \| directory /Users/solomon/my/src` | ❌ |
| Write host filesystem | 😈  | 😈  `export /foo/bar` | 😈  | `export /foo/bar` | ❌ |
| Read git context | 😈  | 😈 | 😈 | 😇 `+defaultPath="/README.md"` *note: for contextual modules only* |
| Read host env variables | 😈 | 😈 `--token env:TOKEN` | ❌ | ❌ |

**Comment by shykes** [comment]
FYI @aluzzardi @jedevc @sipsma and anyone who might be interested in this topic

**Comment by idlsoft** [comment]
host-side scripting is exactly what I'm missing in my setup.
I currently use bash scripts to prepend initialization functions to a dagger call, but this is very limited.
Ideally I'd like to be able to define "host-side" functions, that glue things together.
For example, a `dev-generate` function that drops generated sources to local directories, or a `docker-build` function that exports a docker container to the local engine.
It may be a bit much to do that with a bash-like shell though, Starlark may be a better fit.

**Comment by shykes** [comment]
@idlsoft I think the bash syntax is a good way to enforce that host-side is for the last mile, and if your scripts start growing too large, it's time to move them to code in a regular sandboxed function - which means that has to be a good experience, too.

The second you want your host-side script to be callable by another script (something which is implied with "host-side scripts" and a more structured language like starlark), then you're on the wrong track.

To address your specific use cases:

- For codegen, we need a better way for dagger modules to implement that, no question.
- For access to the local docker engine: I think that's fine as a last-mile script on the client, for example installing a docker compat module, and passing the local engine's unix socket. Using the local docker engine is kind of a stopgap anyway. As we reach feature parity you'll find yourself needing it less and less IMO.

**Comment by idlsoft** [comment]
> The second you want your host-side script to be callable by another script (something which is implied with "host-side scripts" and a more structured language like starlark), then you're on the wrong track.

Not exactly what I had in mind, but I get that this may be well outside the scope here.

I have a bunch of projects, different languages, but all of them should be able to do `build`, `test`, `lint`.
If certain env variables are defined, also `deploy`.

A "host side" skylark script would be like a `Makefile` translating "standard" commands into dagger function calls, a prologue if you like.

Eventually I'd like to be able to build multiple projects within one dagger session and replace an entire github workflow. 
Have dagger take care of running things in parallel, caching, etc.
Moving this inside a function would not be practical because it would mean copying the entire monorepo into the dagger context. It's also not clear how you would load different modules for different projects when you're already inside a function.

Anyway, this is outside the scope here, but perhaps if dagger cli was available as a library it would make it easier for others to experiment.
